src/textual/app.py

```diff
@@ -947,6 +954,32 @@ async def run_app() -> None:
             event_loop.run_until_complete(run_app())
         return self.return_value
 
+    @contextmanager
+    def suspend(self) -> Iterator[None]:
+        """
+        A context manager that temporarily suspends the app while inside the managed block.
+
+        While inside the `with` block, the app will stop reading input and emitting output.
+        Other applications will have full control of the terminal,
+        configured as it was before the app started running.
+
+        When the `with` block ends, the application will start reading input and emitting output again.
+        """
+        driver = self._driver
+        if driver is not None:
+            driver.stop_application_mode()
+            with redirect_stdout(sys.__stdout__), redirect_stderr(sys.__stderr__):
+                # This sleep seems to help avoid timing problems.
+                # Without it, if you do something like
+                #
+                #   with app.suspend():
+                #       print("hi")
+                #
+                # The print will sometimes not go to sys.__stdout__ as expected.
+                time.sleep(0.025)
```

Comment:
This may be masking a race condition. I wonder if we need to do a flush after stopping application mode so no escape sequences arrive after the yield.

Reply:
🤦🏻‍♂️  of course! I will experiment.

Reply:
Initial results: `self.console.file.flush()` before `yield` *does not* fix the problem. Given that it's random whether the prints appear or not, the sleep is definitely masking a race condition, but I'm not sure where it is.

But here's where it gets extra weird: I tried adding some debug code to `_NullFile.write` to see if the write were still going there, but it seems like they aren't, which makes it seem like they're just completely vanishing, which doesn't make much sense. The investigation continues...

Reply:
Curious! Do you think `"hi!"` is not being written at all?

We have a thread that writes to stdout, so that we can prepare the next frame while the previous frame is being printed. I wonder if that thread is still writing in the background for a fraction of a second.

Reply:
Aha! That's promising - I didn't realize that `self.console.file` was that thread (should have `print`ed it!) https://github.com/Textualize/textual/blob/0be1d96107adffcec748dcd89e04be0422ff61b0/src/textual/app.py#L284-L299

And `_WriterThread.flush` is a no-op! https://github.com/Textualize/textual/blob/0be1d96107adffcec748dcd89e04be0422ff61b0/src/textual/app.py#L190-L192

I bet the problem lies in there - I'll try to find some time to investigate today.